retrieving files from completed calculation - file missing

While trying to move the `aiida-zeopp` to aiida 1.0, I have encountered a strange issue when  retrieving files back from a completed calculation using my plugin. 

I'm running the calculation with `run_get_node` and when the calculation finishes, it seems to me like this is happening (see logs below, I've added some extra prints here and there):
 * the first time it tries to retrieve the data, all output files are there except for one (`out.gridG`). I don't know why (I'm running a serial executable).
 * around 20s later, for some reason AiiDA tries again to retrieve the data; then all files are there

Is it possible that AiiDA's transport is somehow too quick for the file system?

I'm running on my local laptop using local transport and direct scheduler

```
03/26/2019 03:17:40 PM <6091> aiida.execmanager: [WARNING] Retrieving calc 85
03/26/2019 03:17:40 PM <6091> aiida.execmanager: [WARNING] try printing workdir
03/26/2019 03:17:40 PM <6091> aiida.execmanager: [WARNING] /private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23
03/26/2019 03:17:40 PM <6091> aiida.execmanager: [WARNING] ['/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.cssr', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.res', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/_scheduler-stderr.txt', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.volpo', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/MgO.cif', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.chan', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/MgO.rad', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/_aiidasubmit.sh', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.vsa', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/_scheduler-stdout.txt', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.strinfo', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.sa']
03/26/2019 03:17:40 PM <6091> aiida.execmanager: [DEBUG] [retrieval of calc 85] chdir /private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] inside retrieving
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] ['out.cssr', 'out.res', 'out.sa', 'out.vsa', 'out.volpo', 'out.chan', 'out.strinfo', 'out.gridG', '_scheduler-stdout.txt', '_scheduler-stderr.txt']
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.cssr
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.cssr'
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.res
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.res'
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.sa
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.sa'
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.vsa
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.vsa'
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.volpo
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.volpo'
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.chan
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.chan'
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.strinfo
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.strinfo'
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmpykhgps44/out.gridG
03/26/2019 03:17:40 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.gridG'
03/26/2019 03:17:40 PM <6091> aiida.engine.transports: [ERROR] Exception whilst using transport:
Traceback (most recent call last):
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/transports.py", line 106, in request_transport
    yield transport_request.future
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/processes/calcjobs/tasks.py", line 238, in do_retrieve
    raise Return(execmanager.retrieve_calculation(node, transport, retrieved_temporary_folder))
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/daemon/execmanager.py", line 267, in retrieve_calculation
    retrieve_files_from_list(calculation, transport, folder.abspath, retrieve_list)
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/daemon/execmanager.py", line 469, in retrieve_files_from_list
    transport.get(rem, os.path.join(folder, loc))
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/transports/plugins/local.py", line 474, in get
    raise IOError("The remote path {} does not exist".format(remotepath))
OSError: The remote path out.gridG does not exist

03/26/2019 03:17:40 PM <6091> aiida.engine.transports: [DEBUG] Transport request closing transport for AuthInfo for leopold.talirz@epfl.ch on localhost
03/26/2019 03:17:40 PM <6091> aiida.orm.nodes.process.calculation.calcjob.CalcJobNode: [ERROR] iteration 1 of do_retrieve excepted, retrying after 20 seconds
Traceback (most recent call last):
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/utils.py", line 175, in exponential_backoff_retry
    result = yield coro()
  File "/Users/leopold/Applications/miniconda3/envs/aiida_zeopp_py3/lib/python3.6/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/Users/leopold/Applications/miniconda3/envs/aiida_zeopp_py3/lib/python3.6/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/Users/leopold/Applications/miniconda3/envs/aiida_zeopp_py3/lib/python3.6/site-packages/tornado/gen.py", line 1069, in run
    yielded = self.gen.send(value)
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/processes/calcjobs/tasks.py", line 238, in do_retrieve
    raise Return(execmanager.retrieve_calculation(node, transport, retrieved_temporary_folder))
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/daemon/execmanager.py", line 267, in retrieve_calculation
    retrieve_files_from_list(calculation, transport, folder.abspath, retrieve_list)
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/daemon/execmanager.py", line 469, in retrieve_files_from_list
    transport.get(rem, os.path.join(folder, loc))
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/transports/plugins/local.py", line 474, in get
    raise IOError("The remote path {} does not exist".format(remotepath))
OSError: The remote path out.gridG does not exist
03/26/2019 03:18:02 PM <6091> aiida.engine.transports: [DEBUG] Transport request opening transport for AuthInfo for leopold.talirz@epfl.ch on localhost
03/26/2019 03:18:02 PM <6091> aiida.engine.processes.calcjobs.tasks: [INFO] retrieving CalcJob<85>
03/26/2019 03:18:02 PM <6091> aiida.execmanager: [WARNING] Retrieving calc 85
03/26/2019 03:18:02 PM <6091> aiida.execmanager: [WARNING] try printing workdir
03/26/2019 03:18:02 PM <6091> aiida.execmanager: [WARNING] /private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23
03/26/2019 03:18:02 PM <6091> aiida.execmanager: [WARNING] ['/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.cssr', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.res', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/_scheduler-stderr.txt', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.volpo', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/MgO.cif', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.chan', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.gridG', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/MgO.rad', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/_aiidasubmit.sh', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.vsa', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/_scheduler-stdout.txt', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.strinfo', '/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23/out.sa']
03/26/2019 03:18:02 PM <6091> aiida.execmanager: [DEBUG] [retrieval of calc 85] chdir /private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/tmpbi_z_qlq/fe/b9/b75b-9260-4bb4-82cf-239fd77b9b23
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] inside retrieving
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] ['out.cssr', 'out.res', 'out.sa', 'out.vsa', 'out.volpo', 'out.chan', 'out.strinfo', 'out.gridG', '_scheduler-stdout.txt', '_scheduler-stderr.txt']
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.cssr
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.cssr'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.res
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.res'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.sa
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.sa'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.vsa
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.vsa'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.volpo
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.volpo'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.chan
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.chan'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.strinfo
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.strinfo'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/out.gridG
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item 'out.gridG'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/_scheduler-stdout.txt
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item '_scheduler-stdout.txt'
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/sandbox/tmp3rov2uv1/_scheduler-stderr.txt
03/26/2019 03:18:02 PM <6091> aiida.transport.LocalTransport: [WARNING] [retrieval of calc 85] Trying to retrieve remote item '_scheduler-stderr.txt'
03/26/2019 03:18:02 PM <6091> aiida.execmanager: [DEBUG] [retrieval of calc 85] Storing retrieved_files=None
03/26/2019 03:18:02 PM <6091> aiida.engine.transports: [DEBUG] Transport request closing transport for AuthInfo for leopold.talirz@epfl.ch on localhost
03/26/2019 03:18:02 PM <6091> aiida.engine.processes.calcjobs.tasks: [INFO] retrieving CalcJob<85> successful
03/26/2019 03:18:02 PM <6091> aiida.engine.persistence: [DEBUG] Persisting process<85>
03/26/2019 03:18:02 PM <6091> aiida.orm.nodes.process.calculation.calcjob.CalcJobNode: [DEBUG] Broadcast message 'None' received with communicator '<kiwipy.rmq.communicator.RmqSubscriber object at 0x117e53908>'
03/26/2019 03:18:02 PM <6091> aiida.execmanager: [DEBUG] [parsing of calc 85] Content of the retrieved_temporary_folder:

03/26/2019 03:18:02 PM <6091> aiida.parser.NetworkParser: [ERROR] ['_scheduler-stderr.txt', '_scheduler-stdout.txt', 'out.chan', 'out.cssr', 'out.gridG', 'out.res', 'out.sa', 'out.strinfo', 'out.volpo', 'out.vsa']
03/26/2019 03:18:02 PM <6091> aiida.parser.NetworkParser: [ERROR] /Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path
03/26/2019 03:18:02 PM <6091> aiida.parser.NetworkParser: [ERROR] retrieve list
03/26/2019 03:18:02 PM <6091> aiida.parser.NetworkParser: [ERROR] ['out.cssr', 'out.res', 'out.sa', 'out.vsa', 'out.volpo', 'out.chan', 'out.strinfo', 'out.gridG', '_scheduler-stdout.txt', '_scheduler-stderr.txt']
03/26/2019 03:18:02 PM <6091> aiida.parser.NetworkParser: [ERROR] ['/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.cssr', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.res', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/_scheduler-stderr.txt', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.volpo', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.chan', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.gridG', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.vsa', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/_scheduler-stdout.txt', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.strinfo', '/Users/leopold/Personal/Postdoc-MARVEL/aiida_folders/aiida_rmq/.aiida/repository/plugin_migration/repository/node/2c/fa/659c-3ad3-4ad4-8d06-3bf0816d56ae/path/out.sa']
03/26/2019 03:18:03 PM <6091> aiida.orm.nodes.process.calculation.calcjob.CalcJobNode: [DEBUG] Broadcast message 'None' received with communicator '<kiwipy.rmq.communicator.RmqSubscriber object at 0x117e53908>'
```

A few more notes:
 * [the tests pass on travis](https://travis-ci.org/ltalirz/aiida-zeopp/jobs/516165043). not sure whether one can deduce that the problem does not exist on ubuntu in general (travis uses singularity + things might be slow anyhow...)
 * > is it possible that zeo++ uses threads or other weird things like spawning another process to write?
    
   I have looked through the code and I don't think so.
 * > maybe change the direct scheduler to treat zombies as RUNNING
   
   I tried but it doesn't fix the issue

 * > Could you add some additional debug output here to see both the raw output of `ps -o pid,stat,user,time` (parameters retail, stdout, stderr) and the parsed output (returned object), to see if for any reason it's thinking the job is done when it's not?

   I did put the debug there and the lists coming from `ps` and the parsed `job_list` have the same length and all processes are considered `RUNNING`. I'm not quite sure what you meant by "parameters retail, stdout, stderr".


 * > Also, does the parser see abc? (just wondering if the file is written later, with an older filestamp)

   Very good point. The answer is no - I put it first in the list of files to be retrieved and:
```
04/06/2019 11:20:53 AM <94501> aiida.transport.LocalTransport: [WARNING] [retrieve list '['abc', 'out.gridG', '_scheduler-stdout.txt', '_scheduler-stderr.txt']'
04/06/2019 11:20:53 AM <94501> aiida.transport.LocalTransport: [DEBUG] [retrieval of calc 475] Trying to retrieve remote item 'abc'
04/06/2019 11:20:53 AM <94501> aiida.engine.transports: [ERROR] Exception whilst using transport:
Traceback (most recent call last):
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/transports.py", line 106, in request_transport
    yield transport_request.future
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/processes/calcjobs/tasks.py", line 238, in do_retrieve
    raise Return(execmanager.retrieve_calculation(node, transport, retrieved_temporary_folder))
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/daemon/execmanager.py", line 260, in retrieve_calculation
    retrieve_files_from_list(calculation, transport, folder.abspath, retrieve_list)
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/engine/daemon/execmanager.py", line 470, in retrieve_files_from_list
    transport.get(rem, os.path.join(folder, loc), ignore_nonexisting=False)
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/transports/plugins/local.py", line 474, in get
    raise IOError("The remote path {} does not exist".format(remotepath))
OSError: The remote path abc does not exist
```
(I did check and the file `abc` *is* there)

So, the next thing I did was replace `touch abc` by `ls -latr --time-style=full-iso > abc` 
and --surprise-- `abc` was empty!
The reason: AiiDA was taking /bin/ls of MacOS while I am using the GNU version /opt/local/libexec/gnubin/ls from macports.

I then went for `"append_text": 'echo "abc" > abc; ls -ltr >> abc; echo "cde" >> abc'` and the content of abc is
```
abc
total 1216
-rw-r--r--  1 leopold  staff     258 Apr  6 12:00 _aiidasubmit.sh
-rw-r--r--  1 leopold  staff     885 Apr  6 12:00 MgO.cif
-rw-r--r--  1 leopold  staff       0 Apr  6 12:00 _scheduler-stderr.txt
-rw-r--r--  1 leopold  staff  604488 Apr  6 12:00 out.gridG
-rw-r--r--  1 leopold  staff     708 Apr  6 12:00 _scheduler-stdout.txt
-rw-r--r--  1 leopold  staff       4 Apr  6 12:00 abc
cde
```
(while the retrieval of the file still fails).

So, it seems clear that:
 * when the execution of the `_aiida_submit.sh` script finishes, all required files are there
 * for some, still undetermined reason, AiiDA does not see them

P.S. I wondered whether this mixture of macos and gnu binaries was somehow responsible for the issue, so I removed the corresponding settings from the bashrc, opened a new terminal and ran the job again - same results.

--- a/aiida/schedulers/plugins/direct.py
+++ b/aiida/schedulers/plugins/direct.py
@@ -93,7 +93,10 @@ class DirectScheduler(aiida.schedulers.Scheduler):
         TODO: in the case of job arrays, decide what to do (i.e., if we want
               to pass the -t options to list each subjob).
         """
-        # x:  include processes which do not have a controlling terminal
+        # Using subprocess.Popen with `preexec_fn=os.setsid` (as done in both local and ssh transport) results in
+        # processes without a controlling terminal.
+        # The -x option tells ps to include processes which do not have a controlling terminal, which would not be
+        # listed otherwise (leading the direct scheduler to conclude that the process already completed).
         command = 'ps -xo pid,stat,user,time'
 
         if jobs:

--- a/aiida/schedulers/plugins/direct.py
+++ b/aiida/schedulers/plugins/direct.py
@@ -93,7 +93,8 @@ class DirectScheduler(aiida.schedulers.Scheduler):
         TODO: in the case of job arrays, decide what to do (i.e., if we want
               to pass the -t options to list each subjob).
         """
-        command = 'ps -o pid,stat,user,time'
+        # x:  include processes which do not have a controlling terminal
+        command = 'ps -xo pid,stat,user,time'
 
         if jobs:
             if isinstance(jobs, six.string_types):

--- a/docs/source/developer_guide/core/internals.rst
+++ b/docs/source/developer_guide/core/internals.rst
@@ -460,3 +460,8 @@ You should be aware of this while developing code which runs in the daemon. In p
 
     print(subprocess.check_output('sleep 3; echo bar', preexec_fn=os.setsid))
 
+.. note::
+
+    When dropping python 2.7 support, ``preexec_fn=os.setsid`` should be replaced
+    by the thread safe ``start_new_session=True`` introduced in python 3.2.
+

